### DISCUSSION TITLE
Use a zapper?

### DISCUSSION THREAD

**Discussion by kndw** [fhoedemakers/pico-infonesPlus#69] [Ideas]
Dear fhoedemakers - thank you so much for this project. I built it and brought it to work. Needless to say, the office has become less productive.

A colleague brought me an original zapper to play around with. I also found [https://neslcdmod.com/roms/](url), which modifies for instance duck hunt to work with a modern display. I would like to try to make this work. Could you give me some pointer on how to add an additional type of controller, such that I can try to incorporate the zapper?

**Comment by fhoedemakers** [comment]
Thanks.

I looked into this. First of all, i don't think it will work with the original NES zapper. You will need a clone.

Then the now unused D3/D4 pins of the nes controller port need to be connected to two available GPIO ports and the emulator software needs to be changed so it can read this two ports. https://www.nesdev.org/zapper_to_famicom.txt

But also, this is no guarantee it will work.

Frank

**Reply by kndw** [reply]
Thanks for your response.

I gathered that there may be issues with using an original zapper, but there appear to be solutions (removing a capacitor), so I might try that.

However, I'm not very knowlegeable programming a pico, and I can't find where to set pin definitions. Could you point me to those?

**Reply by fhoedemakers** [reply]
I will be happy to help you out and program the required GPIO ports into the emulator. (Providing there are still some free ones available)
But I'm recovering from surgery now and still in the hospital, so give me a few weeks to come back to this.

Frank

**Comment by kndw** [comment]
Let's get in touch then. All the best with your recovery!

**Comment by fhoedemakers** [comment]
Just did some tests with a clone Tomee light gun.  Got it somewhat working in the IPS patched Duck Hunt with the highest delay 9.9 (Use in Duck Hunt menu  up to change whole digit and then right to change the decimal digit) I also made the room made very dark.

I tested with a 55inch Samsung TV with the zapper about one meter from the panel.

I don't know what configuration you use, but i tested using the breadboard setup with GPIO 27 connected to D4 and GPIO 28 to D3 **on the second controller port**

**You can only play with the zapper connected to controller port 2.**

You also need to switch off of the scanlines in the emulator. (Press Select + Down repeatedly until scanlines are gone)

Tell me what build you are using so i can attach an .uf2 for for you to test.  (If you are using the Pimoroni DV Demo board, there might be a problem finding free GPIO ports) 

I feel it is still somewhat of a hit and miss. The changed code is in the zapper branch

![image](https://github.com/user-attachments/assets/e608e94d-135b-4a5a-8dc1-d5a37d39baf4)

**Comment by kndw** [comment]
Awesome, I didn't expect such a quick response. I take you your recovery
went well?

I am using the build with the sd and dvi breakout boards. It should not be
a problem to add the pins.

Thank you, and I will report back how it goes!

Best,

Ksander

**Reply by fhoedemakers** [reply]
Still in recovery and need a lot of rest. Did not spend a lot of time for this. I noticed my trial and error programming broke regular two player games. Will fix this the next couple of days.

Do you need a prebuilt .uf2 of the zapper branch or can you build it yourself? Although two player is broken for now, you can then still test the zapper. If you need an .uf2, give me some time to do this.

**Comment by kndw** [comment]
A .uf2 would be nice, since I'm not comfortable with vscode. However, I'm
in no rush -still working on the board and an adapter socket for the
lightgun so please take your time!

**Comment by fhoedemakers** [comment]
[piconesPlus.zip](https://github.com/user-attachments/files/17254663/piconesPlus.zip)

The link above contains the build for the version with SD and DVI breakout boards.
As for now, only "Game A 1 Duck" in Duck Hunt, and Wild Gunman seem to work properly. Games where you have to shoot more than one object do not work.(at least on my LCD TV)
Have to figure out why. 

Non-zapper two player games also work again with a NES controller on port 2.

Also:

- Make sure you ips patch the correct version of Duck Hunt. Do a Google Search on "duck hunt world rom". More info on [https://neslcdmod.com/](https://neslcdmod.com/)
- Zapper must be connected to NES controller port 2.
- Set the delay to 9.9 in the title screen
- Disable scanlines in emulator. (Select + down)
- Play in a moderate lighted room
- The original NES zapper will not work. You need to do a [hardware modification](https://x.com/eNuffGs/status/1046244128365805570) on this, or use a third party zapper [I used the Tomee light gun](https://www.amazon.com/Tomee-Zapp-Gun-NES/dp/B000MEA9TQ)

Still needs some work and polishing to be done.

![image](https://github.com/user-attachments/assets/2651b2d8-1679-41b6-9018-e95db74acd2c)

**Comment by kndw** [comment]
Thank you very much. I have made the connections on my build and modded the
gun, but have to wait for a connector which has to come from China. I will
get back to you in about two weeks!

**Comment by fhoedemakers** [comment]
Great. Is it possible for you to test with different delay settings? Interested in the results you get. Speak to you soon.

**Comment by kndw** [comment]
Dear Frank,

I have received and soldered the connector, patched the rom and placed the
uf2 file you sent me on the pico. I have performed one quick test with the
standard delay setting (it is 5, I think)...

It works!

Notes:
- As suggested somewhere on the internet, I have modded the original zapper
by removing the small capacitor (1nF, it looks like) from under the metal
casing around the circuit board in the zapper.
- The delay is quite noticable: I pull the trigger and only see the duck
drop when I get the impression I missed. This probably has nothing to do
with the emulator, but rather the rom/delay(?)

I will try to finetune the delay and let you know if I get an even better
experience. Come to think of it, I did not try missing the ducks yet. But
that's also for later.

Regardless - this is truly awesome. Thank you very much!

Best,

Ksander

**Reply by fhoedemakers** [reply]
> I will try to finetune the delay and let you know if I get an even better experience. Come to think of it, I did not try missing the ducks yet. But that's also for later.


Thanks! Please let me know the results of your further testing

**Comment by kndw** [comment]
It appears I spoke too soon. Pulling the trigger is counted as a hit,
regardless where I point the zapper, and also when I cover the lens.

This seems to be related to the zapper/wiring. Initially, I had GPIO28
connected to the D3/light sensor pin and GPIO27 to D4/trigger pin, as I
gathered from your instructions. Doing so, the game started as expected;
beginning with the start-screen to select game type. However, pulling the
trigger did not do anything. I then changed the leads (GPIO28 to D4 and
GPIO27 to D3), thinking I might had reversed them. After this, the game
started immediately when selecting the rom (oddly bypassing the
start-screen). Now pulling the trigger counted as a hit, therefore I
thought it all worked. But that now turns out not to be the case. I will
put the leads back as they were, and will try to get my hands on another
zapper to rule it out as the problem. Unless the above tells you something
useful?

Best regards,

Ksander

**Reply by fhoedemakers** [reply]
Try to turn off the scan-lines (select + down). Scanlines make the screen darker. Also test in a darkened room. Finally set the delay higher.
I used this lightgun: https://a.co/d/dzXgWNh
This made the game work at least in single duck mode.

**Comment by kndw** [comment]
It is not an issue of the scanlines (those are off) or room I think: the
game does not respond to the zapper at all, also not just the trigger. I
will test another zapper when I can get one.

**Reply by kndw** [reply]
I did some additional testing on the zapper using an oscilloscope: 

*pulling the trigger makes the d4 pin (bottom row of the column with the Vcc pin on top) go from high (3.3V) to low, and then back up to high with a low-pass filter like response.
*the d3 pin (middle row of the column with the Vcc pin on top) is always high. It does not respond to ambient light (darkness/bright light) at all.

This makes me think that the zapper is broken, but I also wonder which of the GPIO the trigger signal should be connected to: is it indeed GPIO27?

**Reply by fhoedemakers** [reply]
Yes, trigger is GPIO27 connected to D4. Pin is set to high and when trigger is pressed it goes to low. See zapper.cpp in the zapper branch. Same for D3. When light detected it goes to low.

**Comment by fhoedemakers** [comment]
The zapper is implemented in the latest release v0.46
